aws/ci_builder.rb

```diff
@@ -0,0 +1,80 @@
+require 'cdo/rake_utils'
+require_relative 'ci_build_metrics'
+module CiBuilder
+  require_relative '../lib/cdo/rake_utils'
+  extend self
+  BUILD_STARTED = 'build-started'.freeze
+
+  def build
+    Dir.chdir(deploy_dir) do
+      return 0 unless build_required?
+
+      FileUtils.touch BUILD_STARTED
+
+      git_update_count = update_repository
+
+      log_build_history(git_update_count)
+
+      display_commit_link
+
+      install_dependencies
+
+      notify_honeybadger
+
+      execute_ci_task
+    end
+  end
+
+  private
+
+  def build_required?
+    RakeUtils.git_updates_available? || File.file?(BUILD_STARTED) || !CDO.daemon
+  end
+
+  def update_repository
+    RakeUtils.git_fetch
+    count = RakeUtils.git_update_count
+    RakeUtils.git_pull if count > 0
+    [1, count].max # I think this is not needed. unless count can be a real number between 0 and 1.
+  end
+
+  def log_build_history(git_update_count)
+    log = `git log --pretty=format:"%h %s (%an)" -n #{git_update_count}`
+    File.write(deploy_dir('rebuild'), log)
+  end
+
+  def display_commit_link
+    git_revision = RakeUtils.git_revision
+    git_revision_short = GitUtils.git_revision_short
```

Comment:
Yes, probably, note that the paths is the reason why I haven't checked this in.

I definitely have issues there. Even though I have test, and ran in adhoc, the revert happened because a path was missing when I checked it in. I am meeting with @cat5inthecradle  to huddle and rubber duck this, because I can write everything but apparently I am stuck with this smaller issue

Reply:
I added a comment explaining how I am gonna work and test the "require" paths.